lib/reports/dynamic.rb

```diff
@@ -0,0 +1,266 @@
+      restriction_objects = []
+      restrictions.each do |r|
+        unless r.is_a? Enumerable
+          raise ArgumentError, "restriction input must be an Enumerable"
+        end
+
+        unless r.flatten.size == 2
+          raise ArgumentError, "restriction input must be 2 elements long"
+        end
+
+        on, key = r.flatten.first.split(".")
+        val = r.flatten.last
+        restriction_objects << Reports::Restriction.new(on: on, key: key, val: val)
+      end
+      restriction_objects
+    end
+
+    # Puts it all together.
+    def run
+      outf header
+      records do |rec|
+        outf rec.to_s
+      end
+    end
+
+    def header
+      @decorations.join("\t")
+    end
+
+    # Runs the query and yields the records
+    def records
+      return enum_for(:records) unless block_given?
+
+      Cluster.where(restrictions_as_hash).no_timeout.to_a.each do |clusterable|
+        clusterable.send(@base).each do |embedded_doc|
+          # This is almost like a mongo aggregate,
+          # where the `Cluster.where(...)`    is the outer `$match`,
+          # the `clusterable.send(@base)` is the `$unwind: @base`,
+          # and the `matching_record?(embedded_doc)` is the inner `$match`.
+          # Almost. And perhaps that would have been the way to go.
+          next unless matching_record?(embedded_doc)
+          # If we passed matching_record?, then the restrictions are satisfied.
+          # Get the fields in @decoration from the embedded doc, for the output.
+          extracted_fields = @decorations.map { |d| embedded_doc.send(d.key) }
+
+          if extracted_fields.empty?
+            # We may be validating to the point where we can't end up here.
+            # Keeping it during development but TODO remove it.
+            raise ArgumentError, "No fields extractable from #{@base} document"
+          end
+
+          yield extracted_fields
+        end
+      end
+    end
+
+    def restrictions_as_hash
+      # from [ {a.b => c}, {d.e => f} ]
+      # to   {  a.b => c,   d.e => f  }
+      @restrictions.map do |r|
+        [r.on_key, r.val]
+      end.to_h
+    end
+
+    # Does the embedded doc under observation match all our @restrictions?
+    def matching_record?(embedded_doc)
+      matches = []
+      @restrictions.each do |restriction|
+        # If a restriction is {holdings.ocn => 5},
+        # then check the embedded doc (a holdings rec),
+        # and see if its restriction.key field (ocn)
+        # equals the restriction val (5)?
+        doc_field_val = embedded_doc.send(get_field(restriction.key))
+        # Add a truth value to array.
+        matches << (doc_field_val == restriction.val)
+      end
+      # For this embedded_doc, as we checked all restrictions,
+      # did we add only trues? If so, the embedded_doc is a match.
+      matches.all?(true)
+    end
+
+    # Ensure that the model class, e.g. Clusterable::XY, has the field Z
+    def validate_model_field(model_class, field)
```

Comment:
Not covered in tests, doesn't appear to be used

Reply:
Correct. The functionality is in `QueryElement.validate!`, this was vestigial and is now removed.